<a href="https://colab.research.google.com/github/Dawa1022/CSC-44800/blob/main/Sherpa_Dawa_AS04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Imports and pip installations
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

import pandas as pd
import numpy as np

## Part 1. Loading the dataset

In [ ]:
data = load_iris()
df = pd.DataFrame(data=data.data, columns=data.feature_names)
df["target"] = data.target

df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [ ]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

### About the dataset

The dataset has 150 total entries with 50 in each of the three classes. The dataset has no missing attributes. It has four numeric attributes that: sepal length (cm), sepal width (cm), petal length (cm) and petal width (cm). The target column, which describes the class of the speices, is called the label and the classes are represented as: 0 (Iris-Setosa), 1 (Iris-Versicolour) and 2 (Iris-Virginica).

## Part 2: Splitting the dataset into train and test

In [ ]:
X = df[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']]
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3: Logistic Regression

In [ ]:
L_reg = LogisticRegression(max_iter=200).fit(X_train, y_train)

sample_x = np.array([[5.4, 3.7, 4, 1.1]])
sample_y = L_reg.predict_proba(sample_x)[0]

print(f"The probability of belonging to class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of belonging to class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of belonging to class Iris-Virginica (2) is: {sample_y[2]:.5f} ")


The probability of belonging to class Iris-Setosa (0) is: 0.10031 
The probability of belonging to class Iris-Versicolour (1) is: 0.88348 
The probability of belonging to class Iris-Virginica (2) is: 0.01621 


In [ ]:
train_score = L_reg.score(X_train, y_train)
test_score = L_reg.score(X_test, y_test)
print(f"Train score = {train_score:.5f} and Test Score = {test_score:.5f}.")

Train score = 0.97778 and Test Score = 1.00000.


The model's accuracy is measured by the score. The train score is very close to one, indicating that the model has a high accuracy score and can accurately predict the majority of the labels. The model's test score is 1, indicating that it consistently generates perfect predictions.

In [ ]:
print(f"Coefficents: \n{L_reg.coef_}\n")
print(f"Intercepts: \n{L_reg.intercept_}")

Coefficents: 
[[-0.42671353  0.9728761  -2.44456411 -1.03191934]
 [ 0.51313854 -0.2238593  -0.21516952 -0.85167281]
 [-0.08642502 -0.74901679  2.65973363  1.88359215]]

Intercepts: 
[  9.50022538   1.91272796 -11.41295334]


## Part 4: Support Vector Machine

In [ ]:
sup_vec = SVC(probability=True, random_state=42).fit(X_train, y_train)

In [ ]:
sample_x = np.array([[5.4, 3.7, 4, 1.1]])
sample_y = sup_vec.predict_proba(sample_x)[0]
print(f"The probability of belonging to class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of belonging to class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of belonging to class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of belonging to class Iris-Setosa (0) is: 0.04469 
The probability of belonging to class Iris-Versicolour (1) is: 0.94453 
The probability of belonging to class Iris-Virginica (2) is: 0.01078 


In [ ]:
train_score = sup_vec.score(X_train, y_train)
test_score = sup_vec.score(X_test, y_test)
print(f"Train score = {train_score:.5f} and Test Score = {test_score:.5f}.")

Train score = 0.97037 and Test Score = 1.00000.


The model's accuracy is measured by the score. The train score is very close to one same as for Logistic Regression, indicating that the model has a high accuracy score and can accurately predict the majority of the labels. The model's test score is 1, indicating that it consistently generates perfect predictions.

## Part 5: Neural Network

In [ ]:
neural_net = MLPClassifier(max_iter=600, random_state=42).fit(X_train, y_train)

In [ ]:
sample_x = np.array([[5.4, 3.7, 4, 1.1]])
sample_y = neural_net.predict_proba(sample_x)[0]
print(f"The probability of belonging to class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of belonging to class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of belonging to class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of belonging to class Iris-Setosa (0) is: 0.03112 
The probability of belonging to class Iris-Versicolour (1) is: 0.96877 
The probability of belonging to class Iris-Virginica (2) is: 0.00011 


In [ ]:
train_score = neural_net.score(X_train, y_train)
test_score = neural_net.score(X_test, y_test)
print(f"Train score = {train_score:.5f} and Test Score = {test_score:.5f}.")

Train score = 0.98519 and Test Score = 1.00000.


The model's accuracy is measured by the score. The train score is very close to one same as the ones before, indicating that the model has a high accuracy score and can accurately predict the majority of the labels. The model's test score is 1, indicating that it consistently generates perfect predictions.

In [ ]:
diff_neural_net = MLPClassifier(max_iter=1000,random_state=42).fit(X_train, y_train)
new_train_score = diff_neural_net.score(X_train, y_train)
new_test_score = diff_neural_net.score(X_test, y_test)
print(f"Train score = {new_train_score:.5f} and Test Score = {new_test_score:.5f}.")

Train score = 0.98519 and Test Score = 1.00000.


## Part 6: K-Nearest Neighbors

In [ ]:
k_nn = KNeighborsClassifier().fit(X_train, y_train)

In [ ]:
sample_x = np.array([[5.4, 3.7, 4, 1.1]])
sample_y = k_nn.predict_proba(sample_x)[0]
print(f"The probability of belonging to class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of belonging to class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of belonging to class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of belonging to class Iris-Setosa (0) is: 0.00000 
The probability of belonging to class Iris-Versicolour (1) is: 1.00000 
The probability of belonging to class Iris-Virginica (2) is: 0.00000 


In [ ]:
train_score = k_nn.score(X_train, y_train)
test_score = k_nn.score(X_test, y_test)
print(f"Train score = {train_score:.5f} and Test Score = {test_score:.5f}.")

Train score = 0.97037 and Test Score = 1.00000.


The model's accuracy is measured by the score. The train score is very close to one same as the ones before, indicating that the model has a high accuracy score and can accurately predict the majority of the labels. The model's test score is 1, indicating that it consistently generates perfect predictions.

## Part 7: Conclusions and takeaways


On the dataset, all of the models turned out to be fairly good. One reason could be because the features have a strong relationship with the label, allowing the models to perform better. Because it had the greatest train and test scores, the Netural Network classifier performed the best on the dataset. One reason could be that the Netural Network classifier generated a more complex model rather than a simple linear model, resulting in a superior classifier. I was really surprised by the fact that all of the models received a score of one. This could be due to the limited size of the test set, with all projected labels matching the genuine labels.